### ISSUE TITLE
HQKAN transformer inference is launch-bound at decode / small-batch (slower than vanilla GPT-2)

### ISSUE THREAD

**Issue by Jim137** [Jim137/qkan#15]
## Problem

At **autoregressive decode** (B=1, T=1) and **short prefill** (B=1, T≤128), HQKAN-GPT2 is **slower than a plain-MLP GPT-2 of the same dimensions**, despite using 45–60% fewer parameters.

The regression worsens with model size:

| scale                  | vanilla (decode)    | HQKAN (decode)      | gap       |
|------------------------|---------------------|---------------------|-----------|
| GPT-2 small (124M / 68M)   | 0.807 ms (1,239 tok/s) | 1.142 ms (876 tok/s)  | **−29% throughput** |
| GPT-2 1B-scale (759M / 307M) | 1.881 ms (532 tok/s)   | 2.203 ms (454 tok/s)  | **−15% throughput** |

On long prefill and batch prefill HQKAN still wins handsomely (**2.0–2.5× faster than vanilla**), so this is strictly a small-shape problem.

## Root cause

HQKAN's MLP is `Linear(d→els) → QKAN([els, els]) → Linear(els→d)` with `els = ⌈log₂(d)⌉` (e.g. 10 for d=768, 11 for d=1536). The QKAN block itself expands into **~8 separate kernel launches** per forward (solver kernel + `(postacts + bias) * pw` + sum + `F.linear` for the base path + final add + the two outer Linears). A 12-layer GPT-2 therefore issues **~100+ kernel launches per decode step**, vs ~70 for vanilla (which has only 2 dense ops per MLP).

Under Nsight profiling at B=1, T=1, the GPU is **idle ~70% of the wall time** — everything is waiting on Python dispatch + `cudaLaunchKernel` overhead (~2 µs each).

Two concrete issues surfaced in investigation:

### 1. CuTe kernels launched on the null stream

```cpp
// csrc/cute_kernels.cu — DISPATCH_FWD macro before the fix
KERNEL<IOT, 32><<<grid, 32, smem>>>(...);       // no stream arg → null stream
```

`torch.cuda.CUDAGraph.capture()` **silently drops** kernels launched on the null stream — they don't even appear in the captured graph. Replaying the graph produces incorrect output (kernel never fires; output buffer holds stale data from capture). We observed max-diff of 0.26 in fp32 HQKAN forward under replay vs eager, with `CUDAGraph warning: The CUDA Graph is empty` emitted.

### 2. `autograd.Function.apply` fires unconditionally under `no_grad()`

```python
# src/qkan/solver/{cute,flash,cutile}.py — before the fix
needs_grad = theta.requires_grad or x.requires_grad   # still True under no_grad!
```

Parameter tensors keep `requires_grad=True` regardless of the current autograd state. So even inside `torch.no_grad()` the solver wraps the kernel in the custom `autograd.Function`, adding ~6–10 µs of Python overhead per forward. Across 12 layers that's **~80–120 µs per decode step** of avoidable overhead.

## Impact

- Users adopting HQKAN for training-efficiency see **worse inference latency than a vanilla baseline** at the most latency-sensitive workload (single-token decode).
- Anyone trying `torch.cuda.graph` on QKAN today gets silently-wrong outputs due to the null-stream bug.
- `torch.compile` can't recover this either — Dynamo can't trace through the pybind11-registered CuTe op.

## Expected behavior

At every inference shape, HQKAN should be **at least as fast as** a parameter-matched vanilla MLP. The launch-bound regime should be solvable with CUDA-graph replay, which replaces N `cudaLaunchKernel` syscalls with a single `cuGraphLaunch`.

## Fix proposed

See PR in [qkan-inference-pr.md](PR_gpt_inference_fix.md). Summary:

1. Launch CuTe kernels on `at::cuda::getCurrentCUDAStream()` so they're captured correctly.
2. Bypass `autograd.Function` under `torch.no_grad()` / `inference_mode()` across all solvers.
3. Introduce `qkan.compile_inference(module)` — a drop-in wrapper that lazily captures a CUDA graph per unique input shape, with automatic fallback to eager on grad-tracking forwards.
4. Introduce `qkan.graph_submodules(model, sample, predicate=…)` for transformer-style models where whole-model capture fails on SDPA's RNG counter — wraps each matching submodule in place.

## Measured fix

After the fix (same bf16 HQKAN-GPT2, `qkan.compile_inference` applied to each MLP):

| scale / shape              | vanilla          | HQKAN eager (before) | **HQKAN+graph (after)** | vs vanilla |
|---------------------------|-----------------:|---------------------:|------------------------:|-----------:|
| **small decode** B=1×1     | 0.807 ms         | 1.142 ms (0.71×)     | **0.706 ms**            | **1.14×**  |
| **small prefill-S** B=1×128| 0.861 ms         | 1.176 ms (0.73×)     | **0.739 ms**            | **1.17×**  |
| **1B decode** B=1×1        | 1.881 ms         | 2.203 ms (0.85×)     | **1.453 ms**            | **1.29×**  |
| **1B prefill-S** B=1×128   | 2.944 ms         | 2.325 ms (1.27×)     | **2.083 ms**            | **1.41×**  |

HQKAN+graph now beats vanilla at **every** shape tested, and the compute-bound regimes (long prefill, batch prefill) retain their existing 1.3–2.5× advantage.

## Environment

- GPU: NVIDIA GeForce RTX 5090 (SM 120)
- PyTorch 2.9.1+cu128
- bf16 forward path, cute solver, pz_encoding ansatz, reps=1
